specs/waku/v2/waku-store.md

```diff
@@ -2,36 +2,94 @@
 title: Waku
 version: 2.0.0-beta1
 status: Draft
-authors: Oskar Thorén <oskar@status.im>, Dean Eigenmann <dean@status.im>
+authors: Oskar Thorén <oskar@status.im>, Dean Eigenmann <dean@status.im>, Sanaz Taheri <sanaz@status.im>
 ---
 
 # Table of Contents
 
+- [Table of Contents](#table-of-contents)
 - [Abstract](#abstract)
+  - [Waku Store Protocol](#waku-store-protocol)
+  - [Pagination](#pagination)
 - [Wire Specification](#wire-specification)
-  * [Protobuf](#protobuf)
+  - [Protobuf](#protobuf)
+    - [Index](#index)
+    - [PagingInfo](#paginginfo)
+    - [HistoryQuery](#historyquery)
+    - [HistoryResponse](#historyresponse)
+    - [HistoryRPC](#historyrpc)
 - [Changelog](#changelog)
 - [Copyright](#copyright)
 
 # Abstract
 
-`WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes.
+`WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes. It also support pagination for more efficient querying of historical messages.
+
+## Pagination
+
+Nodes supporting the store protocol MAY support pagination. When a node supports pagination, a requester node can set the number of `WakuMessage`s to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
+
+To perform pagination, each `WakuMessage` stored at a Waku node running the store protocol is associated with a unique `Index`. Once a `WakuMessage` is received by a node, it will generate the `Index` of the  `WakuMessage` and then saves the (`Index`, `WakuMessage`) together in its local storage. An `Index` encapsulates fields useful for paging, these fields are a `digest` and a `receivedTime`. The `receivedTime` indicates the time at which the `WakuMessage` is received by the store node and the `digest` is the hash of the received `WakuMessage`.  
+
+
+
+When a `HistoryQuery` arrives, it may contain an extra field `pagingInfo` that carries the information required for paging. The `pagingInfo` contains the `pageSize`, `cursor` and the `direction` as defined below:
+
+**pageSize**
+
+- A positive integer indicating the number of `WakuMessage`s to be included in the `HistoryResponse`
+- If the `pageSize` is omitted then a “default page size” shall be used
+- There is an upper limit for the page size
+- If the number of remaining `WakuMessage`s is less than the `pageSize`, then the `pageSize` is adjusted based on the remaining messages
+
+**Cursor**
+
+-  Holds the `Index` of a waku message
+
+**Direction**
+
+- Is a boolean value
```

Comment:
@staheri14 to limit the need of all this explanation, it probably makes it easier to just make direction an enum

Reply:
done